test/requestinterception-experimental.spec.ts

```diff
@@ -0,0 +1,854 @@
+        await page.setRequestInterception(true);
+        page.on('request', (request) => {
+          if (request.url().endsWith('.css'))
+            request.continue(
+              { headers: { ...request.headers(), xaction: 'continue' } },
+              expectedAction === 'continue' ? 1 : 0
+            );
+          else request.continue({}, 0);
+        });
+        page.on('request', (request) => {
+          if (request.url().endsWith('.css'))
+            request.respond(
+              { headers: { xaction: 'respond' } },
+              expectedAction === 'respond' ? 1 : 0
+            );
+          else request.continue({}, 0);
+        });
+        page.on('request', (request) => {
+          if (request.url().endsWith('.css'))
+            request.abort('aborted', expectedAction === 'abort' ? 1 : 0);
+          else request.continue({}, 0);
+        });
+        page.on('response', (response) => {
+          const { xaction } = response.headers();
+          if (response.url().endsWith('.css') && !!xaction)
+            actionResults.push(xaction as ActionResult);
+        });
+        page.on('requestfailed', (request) => {
+          if (request.url().endsWith('.css')) actionResults.push('abort');
+        });
+
+        const response = await (async () => {
+          if (expectedAction === 'continue') {
+            const [serverRequest, response] = await Promise.all([
+              server.waitForRequest('/one-style.css'),
+              page.goto(server.PREFIX + '/one-style.html'),
+            ]);
+            actionResults.push(serverRequest.headers.xaction as ActionResult);
+            return response;
+          } else {
+            return await page.goto(server.PREFIX + '/one-style.html');
+          }
+        })();
+
+        expect(actionResults.length).toBe(1);
+        expect(actionResults[0]).toBe(expectedAction);
+        expect(response.ok()).toBe(true);
+      });
+    }
+
+    it('should intercept', async () => {
+      const { page, server } = getTestState();
+
+      await page.setRequestInterception(true);
+      page.on('request', (request) => {
+        if (utils.isFavicon(request)) {
+          request.continue({}, 0);
+          return;
+        }
+        expect(request.url()).toContain('empty.html');
+        expect(request.headers()['user-agent']).toBeTruthy();
+        expect(request.method()).toBe('GET');
+        expect(request.postData()).toBe(undefined);
+        expect(request.isNavigationRequest()).toBe(true);
+        expect(request.resourceType()).toBe('document');
+        expect(request.frame() === page.mainFrame()).toBe(true);
+        expect(request.frame().url()).toBe('about:blank');
+        request.continue({}, 0);
+      });
+      const response = await page.goto(server.EMPTY_PAGE);
+      expect(response.ok()).toBe(true);
+      expect(response.remoteAddress().port).toBe(server.PORT);
+    });
+    it('should work when POST is redirected with 302', async () => {
```

Comment:
I'm having trouble trying to understand how this test checks if redirection and 302 works. What is the assertion? Could you please add some code comments?

Reply:
I think these tests are just copied from the original `requestinterception.spec.ts` and I added a pull request to add a comment there to explain this tests origin. Suggesting the same comment here:

```suggestion
    // @see https://github.com/puppeteer/puppeteer/pull/3105
    it('should work when POST is redirected with 302', async () => {
```